pydantic_graph/pydantic_graph/graph.py

```diff
@@ -510,3 +505,109 @@ def _infer_name(self, function_frame: types.FrameType | None) -> None:
                     if item is self:
                         self.name = name
                         return
+
+
+class GraphRun(Generic[StateT, DepsT, RunEndT]):
+    """A stateful run of a graph.
+
+    After being entered, can be used like an async generator to listen to / modify nodes as the run is executed.
+    """
+
+    def __init__(
+        self,
+        graph: Graph[StateT, DepsT, RunEndT],
+        first_node: BaseNode[StateT, DepsT, RunEndT],
+        *,
+        history: list[HistoryStep[StateT, RunEndT]],
+        state: StateT,
+        deps: DepsT,
+        auto_instrument: bool,
+        span: LogfireSpan | None = None,
+    ):
+        self.graph = graph
+        self.history = history
+        self.state = state
+        self.deps = deps
+        self._auto_instrument = auto_instrument
+        self._span = span
+
+        self._next_node = first_node
+        self._started: bool = False
+        self._result: End[RunEndT] | None = None
+
+    @property
+    def is_ended(self):
+        return self._result is not None
+
+    @property
+    def result(self) -> RunEndT:
+        if self._result is None:
+            if self._started:
+                raise exceptions.GraphRuntimeError(
+                    'This GraphRun has not yet ended. Continue iterating with `async for` or `GraphRun.next`'
+                    ' to complete the run before accessing the result.'
+                )
+            else:
+                raise exceptions.GraphRuntimeError(
+                    'This GraphRun has not been started. Did you forget to `await` the run?'
+                )
+        return self._result.data
+
+    async def next(
+        self: GraphRun[StateT, DepsT, T], node: BaseNode[StateT, DepsT, T]
+    ) -> BaseNode[StateT, DepsT, T] | End[T]:
+        """Note: this method behaves very similarly to an async generator's `asend` method."""
+        history = self.history
+        state = self.state
+        deps = self.deps
+
+        next_node = await self.graph.next(node, history, state=state, deps=deps, infer_name=False)
+
+        if isinstance(next_node, End):
+            self._result = next_node
+        else:
+            self._next_node = next_node
+        return next_node
+
+    def __await__(self) -> Generator[Any, Any, typing_extensions.Self]:
```

Comment:
```suggestion
    def __await__(self) -> Awaitable[typing_extensions.Self]:
```
I assume you would have if possible, but could we do this?

Reply:
No, and actually I don't think that change is right — the thing returned by `__await__` is **_not_** itself awaitable, it's just a thing that python's async functionality knows how to turn _into_ something awaitable.

(And just to be explicit, changing the type hint in that way does cause a bunch of type errors.)